Feat/45 create real violin plot

HenrikeW opened the pull request:
<!-- Remove sections from this template that are not used -->
## Related issue(s) and PR(s)
This PR closes #45.

<!-- Include below a description of the changes proposed in the pull request -->

## Type of change

<!-- Please delete options that are not relevant -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

## List of changes made

- modifications to the database
- adding API endpoints
- adding page explore-data
- adding form to choose what to show in a violin plot
- creating a violin plot based on user choices

## Still missing / follow up issues
There are some pieces that are still missing but that I think can be solved in follow-up issues:
- The whole page isn't connected to the sidebar menu yet. Instead, you can reach it via the "Explore data" link on the landing page.
- The variable "dataset" is still hardcoded in the frontend. It should fetch information from the `/types` endpoint. This should probably be used in the sidebar menu and then brought to the explore-data page via params.
- Healthy reference samples are so far ignored for the plots.
- The design doesn't follow the sketches yet. 
- The select protein input should become a search field.

## Screenshot of the fix
<img width="1590" height="1536" alt="image" src="https://github.com/user-attachments/assets/ecf88cae-688c-4a5d-89c6-f20dac6d26e3" />

## Testing

### Preparations (loading data)

* Ensure that you have the latest data (currently a set of three CSV files, available via our internal file sharing service).
   * `data_metabolites_PSCID.csv`
   * `data_miRNA_PSCID.csv`
   * `data_proteins_PSCID.csv`

* To prepare for testing in the development environment:
   1. Remove the development database: `rm vol/database/psc-atlas.db`.
   2. Copy the three CSV files into the `vol/uploads` directory.
   3. Build and start the project: `./compose-dev.sh up --build`.
   4. The loading is done when the log mentions both `Processing file: ...` and `Finished processing file: ...` for each of the three CSV files, and the files have been moved into the `vol/uploads/processed` directory.

* To prepare for testing in the production environment:
   1. Drop the persistent Docker volume: `./compose-prod.sh down -v`.
   2. Build and start the project: `./compose-prod up --build`.
   3. Individually, for each of the three CSV files, run `./compose-prod.sh cp {file} proxy:/home/psc-atlas/vol/uploads/` (where `{file}` is the pathname of one of the CSV files).
   4. Wait until the files are noticed (may take 60s) and loaded.

### Testing the interface
- go to `localhost:3320/explore-data`
- choose x-axis variable, legend variable and protein and create a plot
- play around with the choice options to see if everything works well


## Definition of Done checklist

- [x] I have made an effort making the commit history understandable
- [x] I have performed a self-review of my own code and commented any hard-to-understand areas
- [ ] Existing tests are passing and I wrote unit tests for new functionality
- [x] My changes generate no new warnings
- [ ] Make sure the layout and text follow design sketches, if there are any
- [ ] All Acceptance Criteria (AC) Fulfilled and checked in the Related issue:
- [ ] **Code Review Completed:**
  - The code has been reviewed by at least one team member other than its author (adhering to the four-eyes principle).
  - Considerations: PR reviews, presentations to co-workers, pair/mob programming sessions.
  
- [ ] **Documentation Updated (if applicable):**
  - Relevant documentation is current.
  - Considerations: in-code comments, generated documentation, README files, Swagger docs, Postman collections, Confluence pages, etc.
  
- [ ] **Follow-up Backlog Items Created (if applicable):**
  - Necessary follow-up tasks have been identified and added to the backlog.
  - Considerations: next iteration tasks, performance enhancements, visual improvements, refactoring needs, addressing technical debt.

HenrikeW committed:
feat: explore plotting options with hardcoded data. Issue #45

HenrikeW committed:
feat: add basic form to choose plot variables. Issue #45

HenrikeW committed:
feat: add minor form adjustments. Issue #45

andkaha committed:
feat: Endpoints added for sample and variable data

Endpoints:

	* /v1/variable/names [?contains=str]
	* /v1/sample_types

andkaha committed:
feat: Add "limit" parameter to "/v1/variable/names" endpoint

HenrikeW committed:
feat: add request for plot data. Issue #45

andkaha committed:
fix: Values with capital first letter

andkaha committed:
chore: Formatting

andkaha committed:
feat: Add conditions and their valid values

andkaha committed:
chore: Delete old migration file (will be replaced later)

andkaha committed:
fix: Move data ingester's work directories into the "uploads" directory

andkaha committed:
feat: Make it easier to enable echoing of generated SQL statements

andkaha committed:
feat: Make it possible to filter the variables by sample type

Minor model changes to make it more efficient

andkaha committed:
fix: Use correct types

HenrikeW committed:
feat: various minor improvements. Issue #45

andkaha committed:
feat: Add "/v1/sample/conditions?type=XXX" endpoint

... and put some types in "api_types.py" for later reuse.

andkaha committed:
chore: Not needed after all

andkaha committed:
chore: Better typing and comments

andkaha committed:
feat: Store more things as simple strings, implement /measurement endpoint

andkaha committed:
fix: Sort out response formatting per condition value

andkaha committed:
docs: Add data format document

andkaha committed:
feat: Regenerate initial Alembic migration

andkaha committed:
chore(docs): Spelling

HenrikeW committed:
feat: fetch actual data for user input elements. Issue #45

HenrikeW committed:
feat: fetch the real plot data. Issue #45

andkaha committed:
chore: Be smarter with pyproject.toml (no need to copy it)

andkaha committed:
feat: Reimplemented data models, minimum changes to stats models

andkaha committed:
chore: Regenerated Alembic migration

andkaha committed:
feat: Make ingester script also process individual CSV files

andkaha committed:
chore: uv.lock updated

andkaha committed:
feat: All endpoints are now sample endpoints

andkaha committed:
chore: Support new models

andkaha committed:
feat: Implement endpoints on top of new models

andkaha committed:
fix: Ignore the "PSCID" column

andkaha committed:
fix: Handle conditions correctly when requesting measurements

This involves a dynamic number of joins to the "condition" and
"condition_variable" tables, depending on how many conditions are given.

andkaha committed:
fix: Proper grouping of measurements

HenrikeW committed:
feat: add function to create plot data. Issue #45

HenrikeW committed:
feat: ignore NA, fix checkbox bug, improve plot visuals. Issue #45

andkaha commented on frontend/src/app/explore-data/page.tsx:
```diff
@@ -1,63 +1,382 @@
+              title: {
+                text: "",
+              },
+            }}
```
The `violinmode` option is formally not valid in the layout (even though it makes the plot look the way we expect), which causes the TypeScript compiler to fail the build in production mode.

After looking into this, the only _easy_ way to fix it is to change `}}` on line 337 to `} as any}`.  This is a hack; if I understand it correctly, it effectively bypasses the typechecking for this structure.

HenrikeW replied:
I think it's okay to do it the easy, hacky way here.

pdmnyberg replied:
I would prefer if we avoid removing the type checking. And generally I think that our project are configured to specifically disallow using `any` as a type.

In this case I would suggest defining a new variable as follows: 

```ts
const layout: Partial<Plotly.Layout> & {
    violinmode: "group" | "overlay";
  } = {
    width: 700,
    height: 500,
    violinmode: "overlay",
    title: {
      text: "",
    },
  }
```

And then pass that variable as an argument to layout in the component. Since the type will be considered compatible with `Partial<Plotly.Layout>` which is the expected type from the perspective of the component. 

We have a diff in the version for the `"@types/plotly.js": "^3.0.8",` and `"plotly.js": "^3.2.0",` which I assume may case this deviation. Sadly there does not seem to be any better maintained version of types for plotly atm. and the latest version of `@types/plotly.js` used, is the latest one.

I think a hack is the way to go here. But I would prefer if you used something similar to my suggestion above.

andkaha reviewed (changes requested):
See the code comment about `violinmode`.

~Also, the user can currently select the same condition twice, which generates an SQL error ("ambiguous column name"). Let me know if I should fix this (it's easy), or whether you will make it impossible to select the same condition in both dropdowns.~

EDIT: I made a commit that better handles the case where the user selects the same variable for both conditions.

andkaha committed:
fix: Handle the multiple joins to "condition" etc. better

This allows for both (all) conditions to refer to the same condition
variable and values, regardless of whether the frontend code allows for
this or not.

Also, update lock files.

HenrikeW committed:
fix: don't allow same conditions on x-axis and legend. Issue #45

HenrikeW commented:
I've now added a fix that makes it impossible in the frontend to choose the same variable for x-axis and legend.

I've also realized that there is a bug in the current implementation that makes it impossible to create a plot without choosing a legend variable, which I think should be possible. My suggestion is to create a separate issue for this and don't fix it in this PR.

HenrikeW commented:
I've now created follow-up issues: #60 , #61, #62, #63, #64

pdmnyberg commented on frontend/src/app/explore-data/page.tsx:
```diff
@@ -1,63 +1,394 @@
+                        onChange={handleXAxisValues}
+                        className="form-check-input"
+                      />
+                      <label htmlFor={value}>{value}</label>
```
the `htmlFor` should reference the id. In this case "${selectedXaxis.name}-${value}"

HenrikeW replied:
Thanks for noticing! I missed that (here and in another place). Fixed now.

HenrikeW committed:
fix: address comments from review. Issue #45

pdmnyberg commented on frontend/src/app/explore-data/page.tsx:
```diff
@@ -1,63 +1,398 @@
+        throw new Error();
+      }
+      const fetchedConditions: ConditionResponse = await response.json();
+      setConditions(fetchedConditions.conditions);
```
For datafetching such as this I would recommend using SWR. It makes the code very clean, it seems very flexible and include a lot of convenient features. I have tried to use it in the bird-ringing project and it has been working quite nicely. I am not going to require it here so see this as a for future reference:

https://nextjs.org/docs/pages/building-your-application/data-fetching/client-side#client-side-data-fetching-with-swr

HenrikeW replied:
Cool, I didn't know about this but it sounds really good! Will try to remember it for next time.

pdmnyberg commented on frontend/src/app/explore-data/page.tsx:
```diff
@@ -1,63 +1,398 @@
+    try {
+      const response = await fetch(`/api/v1/sample/conditions?type=${dataset}`);
+      if (!response.ok) {
+        throw new Error();
```
The error should contain some kind of description.

HenrikeW replied:
Fixed now.

pdmnyberg commented on backend/src/psc_atlas/engine.py:
```diff
@@ -12,15 +12,17 @@
     f"sqlite:///{Path.home()}/vol/database/psc-atlas.db",
 )
 
+echo = False
```
Should this be configurable from an environment variable?

andkaha replied:
I don't see a real need to add an environment variable to toggle this. I sometimes (and only sometimes) look at the generated SQL when working on the SQLAlchemy models, but never in production. It's easier to change this variable in the code (picked up by hot reload) than to _restart_ the project with a new environment.

pdmnyberg commented on backend/src/psc_atlas/routers/sample.py:
```diff
@@ -0,0 +1,235 @@
+    with get_session() as session:
+        # Build the base query, joining necessary tables and filtering
+        # by sample type and measurement variable name.
+        query = session.query(Measurement.value)
```
I see that you layout this query differently from the other queries. I assume there is no other reason for than to be consistent with the behaviour that follows where the loop operates on the query.

I am referring to row 128, 80 and 90 when it comes to style of query.

andkaha replied:
I had to rework this endpoint several times because I misunderstood the way the data needed to be fetched.

In the end, I realised I had to perform multiple joins on the same table. By then, the layout of the query may very well have diverged from that of the other endpoints. Since the rest of the code needed no further changes, I did not perform a final check to ensure that all code followed the same pattern.

Would you want me to rework the queries?

pdmnyberg replied:
No. I think this is fine. I was just curios. Thanks.

pdmnyberg reviewed (dismissed):
This seems to work as expected and the code looks good. It is a bit much to look through so I may have omitted looking into some details but it looks reasonable to me.

If I understood things correctly the file `example-data.json` is no longer used and should be removed. Other than that I have added some comments where I was curios about some things. Those comments are mostly to satisfy my curiosity and do not need to be addressed for this to be good enough for me.

Good work!

HenrikeW committed:
fix: delete unused file and add error description. Issue #45

andkaha reviewed (approved):
All good! Well done!
